### DISCUSSION TITLE
The implementation of slicing is not as described in the documentation

### DISCUSSION THREAD

**Discussion by yyy33** [winnow-rs/winnow#578] [Ideas]
Hi, `winnow::error::Needed::Size`  of the document said that the variations in bytes, but the Stream realization of the slice in the return to the number of elements as the unit
````rust
impl<'i, T> Stream for &'i [T]
where
    T: Clone + crate::lib::std::fmt::Debug,
{
    #[inline(always)]
    fn offset_at(&self, tokens: usize) -> Result<usize, Needed> {
        if let Some(needed) = tokens.checked_sub(self.len()).and_then(NonZeroUsize::new) {
            Err(Needed::Size(needed))
        } else {
            Ok(tokens)
        }
    }
}

````

**Comment by epage** [comment]
I've tried tweaking the wording in #579.  Does that resolve the problem?

**Reply by yyy33** [reply]
> I've tried tweaking the wording in #579. Does that resolve the problem?

Yes, and is it possible to calculate Needed here based on cnt and tokens, instead of Unknown?
https://docs.rs/winnow/latest/src/winnow/stream/mod.rs.html#827

**Reply by epage** [reply]
Yeah, I think that was a side effect on lack of clarity on whether it might tokens or offsets.

**Reply by epage** [reply]
Hmm, in reviewing this more, the most meaningful / actionable piece of data is offsets, rather than tokens.  For all but `&str`, these are the same thing.  For all but `&[T]`, this is bytes.

Pushed a new change that switches to that.  It preserves the existing `&str` `offset_at` behavior because we don't know how to turn tokens into offsets without content.
